src/TreeSitter/Importing.hs

```diff
@@ -1,135 +1,271 @@
+import           Control.Monad (void)
+import           Control.Monad.IO.Class
+import           Data.ByteString (ByteString)
+import qualified Data.ByteString as B
+import qualified Data.Map as Map
+import qualified Data.Text as Text
+import           Data.Text.Encoding
+import           Foreign.C.String
 import           Foreign.Marshal.Alloc
+import           Foreign.Marshal.Utils
 import           Foreign.Ptr
 import           Foreign.Storable
-import TreeSitter.Cursor as TS
-import TreeSitter.Node as TS
-import TreeSitter.Parser as TS
-import TreeSitter.Tree as TS
-import qualified Data.Text as Text
-import Control.Effect
-import Control.Effect.Reader
-import Control.Monad.IO.Class
-import Data.Text.Encoding
-import qualified Data.ByteString as B
-import Control.Applicative
-import Control.Monad (void)
+import           GHC.Generics
+import           TreeSitter.Cursor as TS
+import           TreeSitter.Language as TS
+import           TreeSitter.Node as TS
+import           TreeSitter.Parser as TS
+import           TreeSitter.Tree as TS
+import           Data.Proxy
+import           Prelude hiding (fail)
 
-data Expression
-      = NumberExpression Number | IdentifierExpression Identifier
-      deriving (Eq, Ord, Show)
+-- Parse source code and produce AST
+parseByteString :: Building t => Ptr TS.Language -> ByteString -> IO (Either String t)
+parseByteString language bytestring = withParser language $ \ parser -> withParseTree parser bytestring $ \ treePtr ->
+  if treePtr == nullPtr then
+    pure (Left "error: didn't get a root node")
+  else
+    withRootNode treePtr $ \ rootPtr ->
+      withCursor (castPtr rootPtr) $ \ cursor ->
+        runM (runFail (runReader cursor (runReader bytestring buildNode)))
 
-data Number = Number
-      deriving (Eq, Ord, Show)
+-- | Building is the process of iterating over tree-sitter’s parse trees using its tree cursor API and producing Haskell ASTs for the relevant nodes.
+--
+--   Datatypes which can be constructed from tree-sitter parse trees may use the default definition of 'buildNode' providing that they have a suitable 'Generic' instance.
+class Building a where
+  buildNode :: (MonadFail m, Carrier sig m, Member (Reader ByteString) sig, Member (Reader (Ptr Cursor)) sig, MonadIO m) => m a
+  default buildNode :: (MonadFail m, Carrier sig m, GBuilding (Rep a), Generic a, Member (Reader ByteString) sig, Member (Reader (Ptr Cursor)) sig, MonadIO m) => m a
+  buildNode = to <$> gbuildNode
 
-data Identifier = Identifier
-      deriving (Eq, Ord, Show)
+  buildEmpty :: MonadFail m => m a
+  buildEmpty = fail "expected a node"
 
+instance Building Text.Text where
+  buildNode = do
+    node <- peekNode
+    case node of
+      Just node' -> do
+        bytestring <- ask
+        let start = fromIntegral (nodeStartByte node')
+            end = fromIntegral (nodeEndByte node')
+        pure (decodeUtf8 (slice start end bytestring))
+      _ -> fail "expected a node for Text"
```

Comment:
Can you clarify this error message a little? I wouldn't know what it meant if I hit it.